src/event_loop.rs

```diff
@@ -0,0 +1,276 @@
+
+/// Epoll event handler to manage and process epoll events for registered file descriptor.
+///
+/// The `VringEpollHandler` structure provides interfaces to:
+/// - add file descriptors to be monitored by the epoll fd
+/// - remove registered file descriptors from the epoll fd
+/// - run the event loop to handle pending events on the epoll fd
+pub struct VringEpollHandler<S: VhostUserBackend<B>, B: Bitmap + 'static> {
+    epoll_file: File,
+    backend: S,
+    vrings: Vec<Vring<GM<B>>>,
+    thread_id: usize,
+    exit_event_fd: Option<EventFd>,
+    exit_event_id: Option<u16>,
+}
+
+impl<S: VhostUserBackend<B>, B: Bitmap + 'static> VringEpollHandler<S, B> {
+    /// Create a `VringEpollHandler` instance.
+    pub(crate) fn new(
+        backend: S,
+        vrings: Vec<Vring<GM<B>>>,
+        thread_id: usize,
+    ) -> VringEpollResult<Self> {
+        let epoll_fd = epoll::create(true).map_err(VringEpollError::EpollCreateFd)?;
+        let epoll_file = unsafe { File::from_raw_fd(epoll_fd) };
+
+        let handler = match backend.exit_event(thread_id) {
+            Some((exit_event_fd, exit_event_id)) => {
+                epoll::ctl(
+                    epoll_file.as_raw_fd(),
+                    epoll::ControlOptions::EPOLL_CTL_ADD,
+                    exit_event_fd.as_raw_fd(),
+                    epoll::Event::new(epoll::Events::EPOLLIN, u64::from(exit_event_id)),
+                )
+                .map_err(VringEpollError::RegisterExitEvent)?;
+
+                VringEpollHandler {
+                    epoll_file,
+                    backend,
+                    vrings,
+                    thread_id,
+                    exit_event_fd: Some(exit_event_fd),
+                    exit_event_id: Some(exit_event_id),
+                }
+            }
+            None => VringEpollHandler {
+                epoll_file,
+                backend,
+                vrings,
+                thread_id,
+                exit_event_fd: None,
+                exit_event_id: None,
+            },
+        };
+
+        Ok(handler)
+    }
+
+    /// Send `exit event` to break the event loop.
+    pub fn send_exit_event(&self) {
+        if let Some(eventfd) = self.exit_event_fd.as_ref() {
+            let _ = eventfd.write(1);
+        }
+    }
+
+    /// Register a custom event only meaningful to the caller into the epoll fd.
+    ///
+    /// When this event is later triggered, and because only the caller knows what to do about
+    /// it, the backend implementation of `handle_event` will be called. This lets entire control
+    /// to the caller about what needs to be done for this special event, without forcing it
+    /// to run its own dedicated epoll loop for it.
+    pub fn register_listener(
+        &self,
+        fd: RawFd,
+        ev_type: epoll::Events,
+        data: u64,
+    ) -> result::Result<(), io::Error> {
+        if data < self.vrings.len() as u64 {
```

Comment:
Why `data` needs to be lower than the number of vrings? This seems to break `set_vring_kick` here.

Reply:
`data` is used as "event index",  and value lower than self.vrings.len() are reserved for vrings. Please refer to event_loop.rs:214.
```
        if (device_event as usize) < self.vrings.len() {
            let vring = &self.vrings[device_event as usize];
        }
```

Reply:
But, in that case, how do event indexes for the vrings get registered into a `VringEpollHandler`?

Reply:
Good point, it's also called by handler.rs. Then we need to update the doc.

Reply:
Fixed